engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/NadelServiceTypeFilterTransform.kt

```diff
@@ -0,0 +1,144 @@
+            typeNamesOwnedByService = typeNamesOwnedByService,
+            fieldObjectTypesOwnedByService = typesOwnedByService,
+            overallField = overallField,
+        )
+    }
+
+    override suspend fun transformField(
+        executionContext: NadelExecutionContext,
+        transformer: NadelQueryTransformer.Continuation,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        field: ExecutableNormalizedField,
+        state: State,
+    ): NadelTransformFieldResult {
+        // Nothing to query if there are no fields, we need to add selection
+        if (state.fieldObjectTypesOwnedByService.isEmpty()) {
+            val objectTypeNames = state.overallField.parent.getFieldDefinitions(executionBlueprint.schema)
+                .asSequence()
+                .flatMap { fieldDef ->
+                    resolveObjectTypes(
+                        executionBlueprint.schema,
+                        type = fieldDef.type,
+                        onNotObjectType = { type ->
+                            error("Unable to resolve to object type: $type")
+                        }
+                    )
+                }
+                .map {
+                    it.name
+                }
+                .filter {
+                    it in state.typeNamesOwnedByService
+                }
+                .toSet()
+                .toList()
+
+            return NadelTransformFieldResult(
+                newField = null,
+                artificialFields = listOf(
+                    newNormalizedField()
+                        .objectTypeNames(objectTypeNames)
+                        .alias(state.aliasHelper.typeNameResultKey)
+                        .fieldName(Introspection.TypeNameMetaFieldDef.name)
+                        .build(),
+                ),
+            )
+        }
+
+        return NadelTransformFieldResult(
+            newField = field
+                .toBuilder()
+                .clearObjectTypesNames()
+                .objectTypeNames(state.fieldObjectTypesOwnedByService)
+                .children(transformer.transform(field.children))
+                .build(),
+        )
+    }
+
+    override suspend fun getResultInstructions(
+        executionContext: NadelExecutionContext,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        overallField: ExecutableNormalizedField,
+        underlyingParentField: ExecutableNormalizedField?,
+        result: ServiceExecutionResult,
+        state: State,
+    ): List<NadelResultInstruction> {
+        return emptyList()
+    }
+
+    private fun getOwnedTypeNames(
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+    ): Set<String> {
+        return service.underlyingSchema
+            .typeMap
+            .values
+            .asSequence()
+            .map {
+                executionBlueprint.getOverallTypeName(service, it.name)
```

Comment:
I am slightly worried about accidentally exposing types that aren't meant to be.

However I convinced myself this can't happen because

- This set of types is only used to cross check that a type is valid
- The final object type name list is a subset of the original i.e. we can't be exposing types that weren't meant to be
- The original object type name list is based off the overall schema i.e. type was exposed

The only thing we need to do is add validation that for shared types that are used, the underlying schema has the exact underlying type.